addons/iap/models/iap_account.py

```diff
@@ -193,32 +197,33 @@ def get_account_id(self, service_name):
         return self.get(service_name).id
 
     @api.model
-    def get_credits_url(self, service_name, base_url='', credit=0, trial=False, account_token=False):
-        """ Called notably by ajax crash manager, buy more widget, partner_autocomplete, sanilmail. """
+    def get_credits_url(self, service_name):
+        """ Called notably by: buy more widget, partner_autocomplete, snailmail, ... """
         dbuuid = self.env['ir.config_parameter'].sudo().get_param('database.uuid')
-        if not base_url:
-            endpoint = iap_tools.iap_get_endpoint(self.env)
-            route = '/iap/1/credit'
-            base_url = url_join(endpoint, route)
-        if not account_token:
-            account_token = self.get(service_name).account_token
+        endpoint = iap_tools.iap_get_endpoint(self.env)
+        route = '/iap/1/credit'
+        base_url = url_join(endpoint, route)
+        account_token = self._hash_iap_token(self.get(service_name).sudo().account_token)
         d = {
             'dbuuid': dbuuid,
             'service_name': service_name,
             'account_token': account_token,
-            'credit': credit,
+            'hashed': 1,
         }
-        if trial:
-            d.update({'trial': trial})
         return '%s?%s' % (base_url, werkzeug.urls.url_encode(d))
 
+    @api.model
+    def _hash_iap_token(self, key):
+        # disregard possible suffix
+        key = (key or '').split('+')[0]
+        if not key:
+            raise UserError(_('The IAP token provided is invalid or empty.'))
+        return hashlib.sha1(key.encode('utf-8')).hexdigest()
+
     def action_buy_credits(self):
         return {
             'type': 'ir.actions.act_url',
-            'url': self.env['iap.account'].get_credits_url(
-                account_token=self.account_token,
-                service_name=self.service_name,
-            ),
```

Comment:
We still need to specify the account token in this case.
If you rely on `get(service_name)` function to get the IAP account, you could end up with a different account than `self`.

E.g. if you have multiple IAP accounts for a service, one belongs to company A, the other belongs to company B and both are in `self.env.companies` then `get(service_name)` will always return the one with the highest ID.
For one of them, you'll get a link to buy credits for the other account.